ssl/quic/quic_rx_depack.c

```diff
@@ -0,0 +1,612 @@
+{
+    while (PACKET_remaining(pkt) > 0) {
+        uint64_t frame_type;
+
+        if (!ossl_quic_wire_peek_frame_header(pkt, &frame_type))
+            return 0;
+
+        switch (frame_type) {
+        case OSSL_QUIC_FRAME_TYPE_PING:
+            if (mode == collect_ack_data)
+                /* This frame makes the packet ACK eliciting */
+                ackm_data->is_ack_eliciting = 1;
+            /* FALLTHRU */
+        case OSSL_QUIC_FRAME_TYPE_PADDING:
+            /*
+             * We ignore this.
+             * For the PING frame, eliciting an ACK is enough.
+             * Since the PING frame is only a header, we simply
+             * skip it.
+             */
+            ossl_quic_wire_skip_frame_header(pkt, &frame_type);
+            break;
+
+        case OSSL_QUIC_FRAME_TYPE_ACK_WITHOUT_ECN:
+        case OSSL_QUIC_FRAME_TYPE_ACK_WITH_ECN: {
+            OSSL_QUIC_FRAME_ACK ack;
+            OSSL_QUIC_ACK_RANGE *ack_ranges;
+            uint64_t total_ranges = 0;
+            int ok = 1;          /* Assume the best */
+
+            if (!ossl_quic_wire_peek_frame_ack_num_ranges(pkt, &total_ranges)
+                || (ack_ranges = OPENSSL_zalloc(sizeof(ack_ranges[0])
+                                                * total_ranges)))
+                return 0;
+
+            ack.ack_ranges = ack_ranges;
+            ack.num_ack_ranges = total_ranges;
+
+            /* QUESTION: where do I get the ack delay exponent? */
+            if (!ossl_quic_wire_decode_frame_ack(pkt,
+                                                 1 /* ACK delay exponent */,
+                                                 &ack, NULL))
+                ok = 0;
+            if (ok
+                && mode == process_frames
+                && !ossl_ackm_on_rx_ack_frame(GET_CONN_ACKM(connection), &ack,
+                                              packet_space, received))
+                ok = 0;
+
+            OPENSSL_free(ack_ranges);
+            if (!ok)
+                return 0;
+            break;
+        }
+
+        case OSSL_QUIC_FRAME_TYPE_RESET_STREAM: {
+            OSSL_QUIC_FRAME_RESET_STREAM frame_data;
+
+            if (!ossl_quic_wire_decode_frame_reset_stream(pkt, &frame_data))
+                return 0;
+
+            switch (mode) {
+            case collect_ack_data:
+                /* This frame makes the packet ACK eliciting */
+                ackm_data->is_ack_eliciting = 1;
+                break;
+            case process_frames: {
+                QUIC_STREAM *stream
+                    = ssl_get_stream(connection, frame_data.stream_id);
+                int stream_type = ssl_get_stream_type(stream);
+
+                ssl_close_stream(stream);
+                if ((stream_type & SSL_STREAM_TYPE_W) != 0) {
+                    OSSL_STATM *statm = GET_CONN_STATM(connection);
+
+                    ssl_close_connection(connection);
+                    ossl_statm_set_error_state(statm, QUIC_STREAM_STATE_ERROR);
```

Comment:
Why not just make calling `ssl_close_connection` a responsibility of `ossl_statm_set_error_state`? I also think that function should probably raise an error too (much like `SSLfatal` does for TLS).

Reply:
How would `ssl_close_connection()` know whether it's being legitimately closed, or closed because of an error?

Reply:
My suggestion was to do it the other way around, i.e. make calling `ssl_close_connection` a responsibility of `ossl_statm_set_error_state`. You can just call `ssl_close_connection` directly for cases where it is being closed not because of an error.

Reply:
Hmmm, and what should I pass then?  Is every instance of `STREAM_STATE_ERROR` a legitimate reason to close the connection?

Reply:
`QUICfatal` it is.

Reply:
Done